src/Socket.cpp

```diff
@@ -34,57 +36,54 @@ void Socket::write(const std::string& buff) {
     write(buff.c_str(), buff.size());
 }
 
-void Socket::read(char* buff, size_t buffsize) {
-    while(buffsize > 0) {
-        ssize_t res = ::read(_fd, buff, buffsize);
-        if(res == 0) {
-            throw std::runtime_error("Socket::read() - client closed connection");
-        } else if(res < 0) {
-            if(errno == EINTR)
-                continue;
-            throw std::runtime_error("Socket::read() - read() failed");
-        }
+void Socket::read(char* buff, const size_t buffsize) {
+    while(_buffer.size() < buffsize)
+        bufferedRead();
 
-        buff += res;
-        buffsize -= res;
-    }
+    strncpy(buff, _buffer.c_str(), buffsize);
+    buff[buffsize] = '\0';
+    _buffer = _buffer.substr(buffsize);
 }
 
 /*
  * Reads a line terminated by \r\n. Returns length of a received string.
- *
- * Current implementation reads bytes one by one. It's probably not the
- * best approach, but it's simple and doesn't require a temporary buffer
- * which should be somehow shifted all the time. So until this method
- * becomes a bottleneck for someone it's good enough.
  */
-size_t Socket::readLine(char* buff, size_t buffsize) {
-    size_t currsize = 0;
-    // TODO: optimize, figure out how many bytes are available using ioctl
-    while(currsize < buffsize) {
-        ssize_t res = ::read(_fd, buff + currsize, 1);
-        if(res <= 0) {
-            if(errno == EINTR)
-                continue;
+size_t Socket::readLine(char* buff, const size_t buffsize) {
+    while(_buffer.find_first_of('\n', 0) == std::string::npos)
```

Comment:
In this case I want to be sure that _buffer has read a string ended with '\n' from socket (see read discussion) and return this string. The end of line is like a message terminator in HTTP headers, so I cannot process it until I read the whole string.